cmd/car/extract.go

```diff
@@ -33,92 +37,117 @@ func ExtractCar(c *cli.Context) error {
 		outputDir = c.Args().First()
 	}
 
-	bs, err := blockstore.OpenReadOnly(c.String("file"))
-	if err != nil {
-		return err
-	}
+	var store storage.ReadableStorage
+	var roots []cid.Cid
 
-	ls := cidlink.DefaultLinkSystem()
-	ls.TrustedStorage = true
-	ls.StorageReadOpener = func(_ ipld.LinkContext, l ipld.Link) (io.Reader, error) {
-		cl, ok := l.(cidlink.Link)
-		if !ok {
-			return nil, fmt.Errorf("not a cidlink")
+	if c.String("file") == "-" {
+		var err error
+		store, roots, err = NewStdinReadStorage(c.App.Reader)
+		if err != nil {
+			return err
+		}
+	} else {
+		carFile, err := os.Open(c.String("file"))
+		if err != nil {
+			return err
 		}
-		blk, err := bs.Get(c.Context, cl.Cid)
+		store, err = carstorage.OpenReadable(carFile)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		return bytes.NewBuffer(blk.RawData()), nil
+		roots = store.(carstorage.ReadableCar).Roots()
 	}
 
-	roots, err := bs.Roots()
+	ls := cidlink.DefaultLinkSystem()
+	ls.TrustedStorage = true
+	ls.SetReadStorage(store)
+
+	path, err := pathSegments(c.String("path"))
 	if err != nil {
 		return err
 	}
 
+	var extractedFiles int
 	for _, root := range roots {
-		if err := extractRoot(c, &ls, root, outputDir); err != nil {
+		count, err := extractRoot(c, &ls, root, outputDir, path)
+		if err != nil {
 			return err
 		}
+		extractedFiles += count
+	}
+	if extractedFiles == 0 {
+		fmt.Fprintf(c.App.ErrWriter, "no files extracted\n")
```

Comment:
should this be considered an error exit code?

Reply:
yep, have made this a non-zero exit case in the latest push